### ISSUE TITLE
[Performance] Batch and bucket alignment inference for faster CPU/GPU execution

### ISSUE THREAD

**Issue by richard880502** [richard880502/whisperX#1]
## Summary

WhisperX ASR already benefits from batched inference through faster-whisper / CTranslate2, but the forced-alignment stage currently runs the alignment model **one segment at a time** in `whisperx/alignment.py`.

The current path is effectively:

```text
for each ASR segment:
    slice waveform
    -> move one segment to device
    -> run Wav2Vec2/CTC model
    -> log_softmax
    -> move emission back to CPU
    -> trellis + backtrack
```

There is already a TODO in the current code:

```python
# TODO: Probably can get some speedup gain with batched inference here
```

This issue proposes implementing **batched alignment inference with duration-aware bucketing**, with the goal of improving alignment throughput on both CUDA GPUs and CPUs while preserving current alignment quality and API compatibility.

---

## Goals

1. Batch Wav2Vec2/CTC alignment-model inference instead of running one segment per forward pass.
2. Group segments with similar durations to minimize padding waste.
3. Support both existing alignment backends:
   - Hugging Face `Wav2Vec2ForCTC`
   - torchaudio pipeline models
4. Preserve original segment ordering and output format.
5. Keep `batch_size=1` as a compatibility/reference path.
6. Improve both GPU utilization/throughput and CPU throughput.
7. Add reproducible benchmarks and correctness tests.

Non-goal for the first implementation: rewriting the CTC trellis/backtrack algorithm. That can be a second optimization after model inference is no longer the dominant bottleneck.

---

## Current bottleneck

The current implementation does model inference inside the per-segment loop:

```python
for sdx, segment in enumerate(transcript):
    ...
    waveform_segment = audio[:, f1:f2]

    with torch.inference_mode():
        if model_type == "torchaudio":
            emissions, _ = model(waveform_segment.to(device), lengths=lengths)
        elif model_type == "huggingface":
            emissions = model(waveform_segment.to(device)).logits

        emissions = torch.log_softmax(emissions, dim=-1)

    emission = emissions[0].cpu().detach()

    trellis = get_trellis(emission, tokens, blank_id)
    path = backtrack(trellis, emission, tokens, blank_id)
```

This creates several inefficiencies:

- many small GPU kernel launches;
- repeated CPU -> GPU transfer per segment;
- repeated GPU -> CPU synchronization per segment;
- poor GPU occupancy for short segments;
- on CPU, repeated small matrix operations instead of larger batched kernels;
- after ASR is optimized, alignment can become a significant fraction of end-to-end latency.

---

# Proposed design

## 1. Separate alignment into two stages

Refactor alignment into:

```text
Stage A: prepare + batch model inference
    transcript segments
        -> validate
        -> slice waveform
        -> collect metadata
        -> bucket by duration
        -> pad batch
        -> run alignment model
        -> split/crop emissions
        -> restore original order

Stage B: existing CTC alignment post-processing
    one emission per original segment
        -> wildcard handling
        -> get_trellis
        -> backtrack
        -> merge_repeats
        -> word/sentence timestamps
```

The first change should avoid rewriting the existing timestamp logic. This makes the optimization lower-risk and easier to compare against the current implementation.

---

## 2. Introduce an internal alignment work item

A small internal structure/dataclass can keep everything required to restore results correctly:

```python
@dataclass
class AlignmentItem:
    index: int
    start: float
    end: float
    waveform: torch.Tensor
    num_samples: int
    text: str
    text_clean: str
```

Important: `index` must always refer to the original transcript position.

After length sorting/bucketing, final emissions must be mapped back by `index` before post-processing/output.

---

## 3. Add batching parameters

Suggested API:

```python
def align(
    transcript,
    model,
    align_model_metadata,
    audio,
    device,
    interpolate_method="nearest",
    return_char_alignments=False,
    print_progress=False,
    combined_progress=False,
    progress_callback=None,
    batch_size=1,
    bucket_by_length=True,
    max_batch_duration=None,
):
```

### Parameter behavior

`batch_size`
- maximum number of segments in one alignment-model forward pass;
- `1` reproduces the current behavior and provides a baseline/reference path.

`bucket_by_length`
- if enabled, group similarly sized waveform segments before batching;
- strongly recommended when `batch_size > 1`.

`max_batch_duration`
- optional safety limit based on the sum of audio duration/samples in one batch;
- useful because a fixed `batch_size=16` can have very different RAM/VRAM requirements for 3-second vs. 30-second segments.

A batch closes when either limit is reached. This gives more predictable behavior on both CPU and GPU than only using a fixed number of segments.

---

## 4. Duration-aware bucketing

Naive batching can waste a large amount of compute.

Bad batch:

```text
3 sec
4 sec
5 sec
29 sec
```

After padding, the model effectively processes approximately:

```text
29 + 29 + 29 + 29 sec
```

for only 41 seconds of real audio.

Instead, group similar lengths:

```text
Bucket A: 3s, 4s, 5s, 6s
Bucket B: 10s, 11s, 13s, 14s
Bucket C: 24s, 26s, 28s, 29s
```

### Suggested first implementation

1. collect valid `AlignmentItem`s;
2. sort item indices by `num_samples`;
3. create batches in sorted order using `batch_size` / `max_batch_duration`;
4. pad only within each batch;
5. save each result using the original `item.index`;
6. run existing post-processing in original transcript order.

This is deterministic and avoids introducing a complicated scheduler.

---

## 5. Padding and valid-length handling

For each batch:

```python
lengths = torch.tensor(
    [item.num_samples for item in batch_items],
    dtype=torch.long,
)

max_len = max(lengths).item()

waveforms = torch.zeros(
    len(batch_items),
    max_len,
    dtype=audio.dtype,
)

for i, item in enumerate(batch_items):
    waveforms[i, :item.num_samples] = item.waveform
```

The existing minimum waveform length of 400 samples must still be respected.

The implementation must also crop model outputs back to each sample's valid emission length so padded frames are never passed into `get_trellis()`.

---

# Backend-specific handling

## 6. torchaudio backend

The torchaudio alignment model already supports batched waveform input and a `lengths` tensor.

Conceptually:

```python
with torch.inference_mode():
    emissions, output_lengths = model(
        waveforms.to(device),
        lengths=lengths.to(device),
    )

    emissions = torch.log_softmax(emissions, dim=-1)
```

Then for every item:

```python
valid = output_lengths[i]
emission_i = emissions[i, :valid]
```

Only valid frames should be copied to CPU.

---

## 7. Hugging Face Wav2Vec2 backend

For Hugging Face models, build an attention mask for padded samples:

```python
attention_mask = torch.arange(max_len)[None, :] < lengths[:, None]
```

Then:

```python
with torch.inference_mode():
    outputs = model(
        waveforms.to(device),
        attention_mask=attention_mask.to(device),
    )
    emissions = torch.log_softmax(outputs.logits, dim=-1)
```

The valid CTC output length must be derived from each waveform length rather than using the padded output length.

For `Wav2Vec2ForCTC`, use the model's feature-extractor output-length calculation where available, e.g.:

```python
output_lengths = model._get_feat_extract_output_lengths(lengths.to(device))
```

This should be wrapped in a helper so backend-specific/private API details stay isolated.

Important: test attention-mask behavior against the existing single-item result for every default alignment model family used by WhisperX.

---

## 8. Avoid per-segment GPU -> CPU synchronization

Current code synchronizes after every segment:

```python
emission = emissions[0].cpu().detach()
```

With batching, prefer:

```text
one model forward
    -> split/crop valid emissions
    -> transfer required emissions to CPU
    -> continue CTC post-processing
```

The first implementation can transfer each cropped emission to CPU after each batch forward pass. A later version can profile whether keeping more post-processing on GPU is worthwhile.

---

# CPU / GPU behavior

## 9. Keep batching device-independent

The batcher should not contain separate CPU and CUDA implementations. The same batching logic should work for both.

Only recommended/default tuning should differ.

Example benchmark candidates:

```text
CUDA:
    batch_size = 1, 4, 8, 16, 32

CPU:
    batch_size = 1, 2, 4, 8
```

Do not hard-code a large CUDA batch size as the global default until measurements are available.

For CPU testing, thread count must also be reported because it strongly affects results:

```python
torch.set_num_threads(N)
```

or the equivalent runtime configuration.

---

## 10. Mixed precision should be separate/optional

After batching is correct, optionally benchmark CUDA autocast:

```python
with torch.inference_mode(), torch.autocast(
    device_type="cuda",
    dtype=torch.float16,  # or bfloat16 where appropriate
):
    ...
```

This should not be required for the first batching PR because it changes numerical behavior and makes correctness comparison harder.

Recommended order:

1. batching + bucketing using current dtype;
2. establish correctness/performance baseline;
3. add optional FP16/BF16 optimization;
4. test timestamp differences separately.

---

# Follow-up optimization after batching

Once model inference is faster, profiling may show `get_trellis()` and `backtrack()` becoming the next bottleneck.

Current CTC post-processing still contains Python/time-step loops.

Possible follow-up work:

- profile `get_trellis` and `backtrack` separately;
- reduce Python overhead;
- avoid unnecessary tensor allocations;
- experiment with compiled/vectorized implementations;
- consider parallel CPU post-processing for independent segments;
- avoid excessive GPU/CPU synchronization.

This should be a separate optimization unless profiling proves it is already dominant.

---

# Implementation plan

## Phase 1 — profiling baseline

Add timing around:

```text
alignment preprocessing
alignment model inference
GPU -> CPU emission transfer
CTC trellis
CTC backtrack
word/sentence post-processing
```

Record the baseline before changing behavior.

## Phase 2 — batch preparation

- extract valid segments into `AlignmentItem`s;
- retain original indices;
- implement duration sorting/bucketing;
- implement padded waveform construction;
- keep `batch_size=1` working.

## Phase 3 — batched model inference

- implement torchaudio batch inference;
- implement Hugging Face batch inference;
- correctly calculate/crop valid emission lengths;
- restore emissions to original segment order.

## Phase 4 — reuse existing alignment logic

Feed each cropped emission into the existing:

```text
wildcard handling
-> get_trellis
-> backtrack
-> merge_repeats
-> timestamp generation
```

Avoid changing this logic in the first PR unless necessary.

## Phase 5 — benchmark and choose defaults

Benchmark CPU/GPU batch sizes and padding efficiency before changing the public default from `1`.

---

# Testing plan

## A. Unit tests — batching behavior

Test at least:

1. `batch_size=1` matches current implementation.
2. `batch_size=2/4` produces the same number of output segments.
3. output segment order is identical to input order even after length sorting.
4. very short waveform (`<400` samples) still works.
5. segments with no alignable characters keep current fallback behavior.
6. segments beyond audio duration keep current fallback behavior.
7. a batch containing very different lengths is cropped correctly.
8. final batch smaller than `batch_size` works.
9. `max_batch_duration` closes batches correctly.
10. empty transcript works.

Run backend-specific tests for both:

```text
model_type == "torchaudio"
model_type == "huggingface"
```

---

## B. Correctness / regression tests

Use the current implementation as the reference.

For the same audio/transcript pair compare:

- segment text;
- segment count/order;
- word count/order;
- word start timestamps;
- word end timestamps;
- alignment scores;
- fallback behavior.

### Suggested timestamp acceptance

Ideally results are identical at the current 3-decimal output precision.

If small floating-point differences appear because of batching, define tolerance based on the alignment model frame resolution. A practical initial regression threshold can be <= 1–2 CTC frames, and any larger difference should be investigated.

Do not judge this optimization using WER alone because the transcript text is already known during forced alignment. Timestamp parity is the important regression metric.

---

## C. Languages / models

At minimum benchmark one model from each current backend path.

Recommended:

### English

```text
torchaudio:
WAV2VEC2_ASR_BASE_960H
```

### Chinese

```text
Hugging Face:
jonatasgrosman/wav2vec2-large-xlsr-53-chinese-zh-cn
```

Chinese is especially useful because the default alignment model is relatively large, so batching may provide a meaningful throughput improvement.

---

# Performance benchmark methodology

## D. Audio workloads

Use multiple workloads rather than one short file:

```text
~5 minute audio   : quick iteration
~30 minute audio  : realistic benchmark
~60 minute audio  : sustained throughput / memory behavior
```

Include both:

- many short VAD/ASR segments;
- mixed short/long segments.

The mixed-length case is necessary to measure whether bucketing is actually reducing padding waste.

---

## E. CPU benchmark

Record:

```text
CPU model
physical/logical core count
RAM
PyTorch version
alignment model
number of torch threads
batch size
bucketing enabled/disabled
```

Test:

```text
batch_size = 1
batch_size = 2
batch_size = 4
batch_size = 8
```

Metrics:

- alignment wall-clock time;
- alignment real-time factor (RTF);
- audio-seconds processed per second;
- CPU utilization;
- peak RSS memory;
- preprocessing time;
- model inference time;
- trellis/backtrack time.

This helps identify where larger batches stop helping because of memory bandwidth/cache/thread contention.

---

## F. GPU benchmark

Record:

```text
GPU model
VRAM
CUDA version
PyTorch version
alignment model
dtype
batch size
bucketing enabled/disabled
```

Test:

```text
batch_size = 1
batch_size = 4
batch_size = 8
batch_size = 16
batch_size = 32  # if memory permits
```

Metrics:

- alignment wall-clock time;
- alignment RTF;
- model-inference time;
- GPU utilization;
- peak VRAM;
- CPU utilization;
- GPU -> CPU transfer time;
- trellis/backtrack time;
- padding ratio.

Suggested monitoring during the test:

```bash
nvidia-smi dmon -s pucm
```

A successful batching implementation should reduce the current saw-tooth pattern caused by many small per-segment forwards and increase sustained GPU utilization.

---

## G. Measure padding efficiency

Add/debug-log these values during benchmarks:

```text
real_audio_samples
padded_audio_samples
padding_ratio = padded_audio_samples / real_audio_samples
```

Compare:

```text
batching without bucketing
vs.
batching with bucketing
```

This validates that bucketing actually improves compute efficiency rather than only changing scheduling.

---

## H. End-to-end benchmark

Alignment-only speedup is necessary, but also report complete WhisperX pipeline timing:

```text
VAD
ASR
alignment
diarization (if enabled)
total
```

Example result table:

| Device | Align batch | Bucketing | Align time | Align RTF | Total pipeline time | Peak RAM/VRAM |
|---|---:|---|---:|---:|---:|---:|
| CPU | 1 | off | baseline | baseline | baseline | ... |
| CPU | 4 | on | ... | ... | ... | ... |
| GPU | 1 | off | baseline | baseline | baseline | ... |
| GPU | 16 | on | ... | ... | ... | ... |

Use the median of at least 3 measured runs after a warm-up run. Report raw results as well as percentage speedup.

---

# Profiling tools

For detailed profiling, useful options include:

```python
torch.profiler
```

and for CUDA:

```bash
nvidia-smi dmon -s pucm
```

The profiling should answer:

1. how much time is currently spent in alignment-model forward passes?
2. how much is GPU/CPU synchronization?
3. after batching, does the bottleneck move to trellis/backtrack?
4. what batch size gives the best throughput before memory/padding overhead dominates?

---

# Acceptance criteria

- [ ] `align(..., batch_size=1)` preserves existing behavior.
- [ ] batched inference works on CPU.
- [ ] batched inference works on CUDA.
- [ ] Hugging Face alignment models work correctly.
- [ ] torchaudio alignment models work correctly.
- [ ] variable-length input is cropped using valid emission lengths.
- [ ] original transcript/segment ordering is preserved.
- [ ] duration bucketing can be enabled/disabled.
- [ ] no meaningful timestamp regression versus the current single-segment path.
- [ ] no regression in existing fallback/error handling.
- [ ] CPU benchmark results are documented.
- [ ] GPU benchmark results are documented.
- [ ] peak RAM/VRAM is measured.
- [ ] padding ratio is measured.
- [ ] end-to-end WhisperX timing is measured in addition to alignment-only timing.

---

## Expected outcome

The main expected improvement is higher alignment-model throughput by replacing many small forward passes with fewer, better-utilized batched forwards.

On GPU this should improve kernel efficiency and utilization. On CPU it should allow larger matrix operations and reduce repeated per-segment invocation overhead. Duration-aware bucketing should prevent padding from erasing those gains.

After this change, profiling should be repeated. If `get_trellis()` / `backtrack()` becomes the dominant cost, that should be handled as a follow-up optimization rather than mixing both changes into the initial batching implementation.
